frontend/app/src/next/components/runs/runs-table/data-table.tsx

```
@@ -0,0 +1,254 @@
+  columns: ColumnDef<TData, TValue>[];
+  data: TData[];
+  emptyState?: React.ReactNode;
+  isLoading?: boolean;
+  selectedTaskId?: string;
+  onRowClick?: (row: TData) => void;
+  onSelectionChange?: (selectedRows: TData[]) => void;
+  rowSelection?: RowSelectionState;
+  setRowSelection?: OnChangeFn<RowSelectionState>;
+  selectAll?: boolean;
+  getSubRows?: (originalRow: TData, index: number) => TData[];
+}
+
+export function DataTable<TData extends IDGetter, TValue>({
+  columns,
+  data,
+  emptyState,
+  isLoading,
+  selectedTaskId,
+  onRowClick = () => {},
+  onSelectionChange,
+  rowSelection = {},
+  setRowSelection,
+  selectAll = false,
+  getSubRows,
+}: DataTableProps<TData, TValue>) {
+  const navigate = useNavigate();
+  const [columnFilters, setColumnFilters] = useState<ColumnFiltersState>([]);
+  const [columnVisibility, setColumnVisibility] = useState<VisibilityState>({});
+  const [expanded, setExpanded] = useState<ExpandedState>({});
+
+  const memoizedRowSelection = useMemo(() => {
+    if (selectAll) {
+      return data.reduce((acc, _, index) => ({ ...acc, [index]: true }), {});
+    }
+    return rowSelection;
+  }, [selectAll, data, rowSelection]);
+
+  const table = useReactTable({
+    data,
+    columns,
+    state: {
+      columnFilters,
+      columnVisibility,
+      rowSelection: memoizedRowSelection,
+      expanded,
+    },
+    onColumnFiltersChange: setColumnFilters,
+    onColumnVisibilityChange: setColumnVisibility,
+    onRowSelectionChange: setRowSelection,
+    enableRowSelection: true,
+    getCoreRowModel: getCoreRowModel(),
+    getSortedRowModel: getSortedRowModel(),
+    getFilteredRowModel: getFilteredRowModel(),
+    getFacetedRowModel: getFacetedRowModel(),
+    getFacetedUniqueValues: getFacetedUniqueValues(),
+    getRowId: (row) => {
+      const typedRow = row as { taskExternalId?: string; id?: string };
+      return typedRow.taskExternalId || typedRow.id || String(Math.random());
```

I think this is potentially a source of bugs in the future - I feel like we should guarantee that `row` always has an `id` or `externalId` or something and definitely not fall back to a random string, but could be convinced otherwise

---

we'll let this one bite us later ;)